File: security/src/test/groovy/io/micronaut/docs/security/authorization/SiteminderAuthenticationFetcher.java
package io.micronaut.docs.security.authorization;

//tag::clazz[]

import io.micronaut.context.annotation.Requires;
import io.micronaut.core.util.StringUtils;
import io.micronaut.http.HttpRequest;
import io.micronaut.security.authentication.Authentication;
import io.micronaut.security.filters.AuthenticationFetcher;
import jakarta.inject.Singleton;
import java.util.Collection;
import java.util.Collections;
import org.reactivestreams.Publisher;
import reactor.core.publisher.Mono;
//end::clazz[]

@Requires(property = "spec.name", value = "SiteminderAuthorizationSpec")
//tag::clazz[]
@Singleton
public class SiteminderAuthenticationFetcher implements AuthenticationFetcher {

    public static final String SITEMINDER_USER_HEADER = "SM_USER";

    @Override
    public Publisher<Authentication> fetchAuthentication(HttpRequest<?> request) {
        return Mono.<Authentication>create(emitter -> {
            String siteminderUser = request.getHeaders().get(SITEMINDER_USER_HEADER);
            if (StringUtils.isEmpty(siteminderUser)) {
                emitter.success();
                return;
            }

            Collection<String> roles = Collections.singleton("ROLE_USER");
            emitter.success(Authentication.build(siteminderUser, roles));
        });
    }
}
//end::clazz[]

File: test-suite-python/src/test/python/micronaut/security/docs/siteminder/SiteminderAuthenticationFetcher.py
# tag::clazz[]
from jakarta.inject import Singleton
from micronaut.context.annotation import Requires
from micronaut.core.util import StringUtils
from micronaut.http import HttpRequest
from micronaut.security.authentication import Authentication
from micronaut.security.filters import AuthenticationFetcher
from org.reactivestreams import Publisher
from reactor.core.publisher import Mono
# end::clazz[]

SITEMINDER_USER_HEADER = "SM_USER"


@Requires(property="spec.name", value="SiteminderAuthorizationTest")
# tag::clazz[]
@Singleton
class SiteminderAuthenticationFetcher(AuthenticationFetcher[HttpRequest]):

    SITEMINDER_USER_HEADER = "SM_USER"

    def fetchAuthentication(self, request: HttpRequest) -> Publisher[Authentication]:
        def fetch(emitter):
            siteminderUser = request.getHeaders().get(self.SITEMINDER_USER_HEADER)
            if StringUtils.isEmpty(siteminderUser):
                emitter.success()
                return

            roles = ["ROLE_USER"]
            emitter.success(Authentication.build(siteminderUser, roles))

        return Mono.create(fetch)
# end::clazz[]
